lib/utils.js

```diff
@@ -39,6 +39,172 @@ function splitReference(reference) {
   };
 }
 
+/**
+ * Determine if a capability is supported by both the provided OAuth scope and capability statement
+ *
+ * @example
+ *
+ * capabilityStatement = {
+ *   "resourceType": "CapabilityStatement",
+ *   "status": "active",
+ *   "date": "2018-03-30T16:04:37+00:00",
+ *   "publisher": "Not provided",
+ *   "kind": "instance",
+ *   "software": {
+ *     "name": "HAPI FHIR Server",
+ *     "version": "2.4"
+ *   },
+ *   "implementation": {
+ *     "description": "HSPC Reference API Server"
+ *   },
+ *   "fhirVersion": "3.0.1",
+ *   "acceptUnknown": "extensions",
+ *   "format": [
+ *     "application/fhir+xml",
+ *     "application/fhir+json"
+ *   ],
+ *   "rest": [
+ *     {
+ *       "mode": "server",
+ *       "resource": [
+ *         {
+ *           "extension": [
+ *             {
+ *               "url": "http:hl7api.sourceforge.net/hapi-fhir/res/extdefs.html#resourceCount",
+ *               "valueDecimal": 1629
+ *             }
+ *           ],
+ *           "type": "Patient",
+ *           "profile": {
+ *             "reference": "http:hl7.org/fhir/Profile/Patient"
+ *           },
+ *           "interaction": [
+ *             {
+ *               "code": "read"
+ *             }
+ *           ]
+ *         }
+ *       ]
+ *     }
+ *   ]
+ * }
+ *
+ * capabilitySupported = scopeSupportFor('patient/Patient.read', 'openid patient/*.*', capabilityStatement)
+ * console.log(capabilitySupported);
+ *
+ * @param {String} capability the scope level, resource, and interaction to check
+ * @param {String} scope the current OAuth scope
+ *
+ * @return {boolean} Returns a determination of support for the given capability in the given scope
+ */
+function supportFor(capability, scope, capabilityStatement) {
+  return scopeSupportFor(capability, scope) && serverSupportFor(capability, capabilityStatement);
+}
+
+/**
+ * Determine if a capability is supported in the given OAuth scope
+ * Scope support will be based on a regex match of any of the following combinations:
+ * level/*.*
+ * level/resource.*
+ * level/resource.interaction
+ * level/*.interaction
+ *
+ * @example
+ *
+ * capabilitySupported = scopeSupportFor('patient/Patient.read', 'openid user/*.* patient/*.*')
+ * console.log(capabilitySupported);
+ *
+ * @param {String} capability the scope level, resource, and interaction to check
+ * @param {String} scope the current OAuth scope
+ *
+ * @return {boolean} Returns a determination of support for the given capability in the given scope
+ */
+function scopeSupportFor(capability, scope) {
+  const [ level, resource, interaction ] = capability.split(/[?:\/\.]/);
+  const permissions = scope.split(' ');
+
+  const permission = permissions.find((permission) => {
+    return permission.match(`${level}\\/(\\*.\\*|${resource}\\.${interaction}|\\*.${interaction}|${resource}\\.\\*)`);;
```

Comment:
This regex is the equivalent of:
```
  const permission = permissions.find((permission) => {
    const exactMatch = permission === capability;
    const levelWildcards = permission === `${level}/*.*`;
    const resourceWildcard = permission === `${level}/*.${interaction}`;
    const interactionWildcard = permission === `${level}/${resource}.*`;

    return exactMatch || levelWildcards || resourceWildcard || interactionWildcard;
  });
```
If the latter is more readable, is it preferable to the regex?